packages/loopring_v3/contracts/amm/libamm/AmmWithdrawal.sol

```diff
@@ -0,0 +1,169 @@
+import "./AmmExchange.sol";
+import "./AmmData.sol";
+import "./AmmPoolToken.sol";
+import "./AmmStatus.sol";
+import "./AmmUtil.sol";
+
+
+/// @title AmmWithdrawal
+library AmmWithdrawal
+{
+    using AmmExchange       for AmmData.State;
+    using AmmPoolToken      for AmmData.State;
+    using AmmStatus         for AmmData.State;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using MathUint96        for uint96;
+    using SafeCast          for uint;
+    using SignatureUtil     for bytes32;
+
+    bytes32 constant public WITHDRAW_TYPEHASH = keccak256(
+        "Withdraw(address joinOwner,uint256 joinOwner,uint256 validUntil)"
+    );
+
+    function withdraw(AmmData.State storage S)
+        public
+    {
+        (uint newJoinLocksStartIdx, uint96[] memory amounts) = getWithdrawableAmounts(S, msg.sender);
+
+        AmmData.TokenLock[] storage joinLocks = S.joinLocks[msg.sender];
+        uint oldJoinLocksStartIdx = S.joinLocksStartIdx[msg.sender];
+        S.joinLocksStartIdx[msg.sender] = newJoinLocksStartIdx;
+
+        for (uint i = oldJoinLocksStartIdx; i < newJoinLocksStartIdx; i++) {
+            delete S.approvedTx[joinLocks[i].txHash];
+            delete joinLocks[i];
+        }
+
+        for (uint i = 0; i < S.tokens.length; i++) {
+            address token = S.tokens[i].addr;
+            delete S.withdrawable[token][msg.sender];
+            AmmUtil.transferOut(token, amounts[i], msg.sender);
+        }
+    }
+
+    function withdrawInShutdown(
+        AmmData.State storage S
+        )
+        public
+    {
+        uint burnAmount = S.balanceOf[msg.sender];
+        require(burnAmount > 0, "INVALID_POOL_AMOUNT");
+
+        IExchangeV3 exchange = S.exchange;
+        bool withdrawalMode = exchange.isInWithdrawalMode();
+
+        uint size = S.tokens.length;
+        for (uint i = 0; i < size; i++) {
+            address token = S.tokens[i].addr;
+
+            require(
+                withdrawalMode && exchange.isWithdrawnInWithdrawalMode(S.accountID, token) ||
+                !withdrawalMode && !exchange.isForcedWithdrawalPending(S.accountID, token),
+                "PENDING_WITHDRAWAL"
+            );
+
+            // Check that nothing is withdrawable anymore.
+            require(
+                exchange.getAmountWithdrawable(address(this), token) == 0,
+                "MORE_TO_WITHDRAWAL"
+            );
+        }
+
+        // Withdraw proportionally to the liquidity owned, including the pool token itself
+        for (uint i = 0; i < size; i++) {
+            address token = S.tokens[i].addr;
+
+            uint balance = token == address(0) ?
+                address(this).balance :
+                ERC20(token).balanceOf(address(this));
+
+            // Withdraw the part owned by the pool
+            uint amount = balance
+                .sub(S.withdrawableBeforeShutdown[token])
+                .mul(burnAmount) / S.totalSupply;
```

Comment:
Currently needs to call `totalSupply()`, but depends on if we always burn all liquidity tokens or not (`poolSupplyToBurn`).